### PR TITLE
Add ClientInstanceTokenBinding type

### PR DESCRIPTION
**Pull request by VivekVinushanth**
### Proposed changes in this pull request
This PR adds new binding type where the client instance can send an identifier of its preference and request the IS to bind itself to the token. This provides the capability for grant types like password or token exchange where session/cookie based binding types cannot be used and also ensures each client instance gets token so that eventually the client has multiple active access token. 

Fixes: https://github.com/wso2/product-is/issues/18600

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/AccessTokenIssuer.java [hunk 1]
```diff
@@ -1065,11 +1066,23 @@ private void handleTokenBinding(OAuth2AccessTokenReqDTO tokenReqDTO, String gran
                 throw new IdentityOAuth2ClientException(OAuth2ErrorCodes.INVALID_REQUEST,
                         "TLS certificate not found in the request.");
             }
+            if (OAuth2Constants.TokenBinderType.CLIENT_INSTANCE.equals(tokenBinder.getBindingType())) {
```
Comment (senthalan) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/AccessTokenIssuer.java:1069]: Having specific checks for implementation in the core is not a good practice. What is the reason for these checks?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/AccessTokenIssuer.java [hunk 2]
```diff
@@ -1065,11 +1066,23 @@ private void handleTokenBinding(OAuth2AccessTokenReqDTO tokenReqDTO, String gran
...
+                tokReqMsgCtx.setTokenBinding(null);
+                return;
+            }
             throw new IdentityOAuth2Exception(
                     "Token binding reference cannot be retrieved form the token binder: " + tokenBinder
                             .getBindingType());
         }
 
+        if (OAuth2Constants.TokenBinderType.CLIENT_INSTANCE.equals(tokenBinder.getBindingType()) &&
+                tokenBindingValueOptional.get().length() >= MAX_ALLOWED_LENGTH) {
```
Comment (senthalan) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/AccessTokenIssuer.java:1080]: can't we do this validation within the ClientInstanceTokenBinder itself ?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/bindings/impl/ClientInstanceTokenBinder.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+public class ClientInstanceTokenBinder extends AbstractTokenBinder {
+
+    private static final String CLIENT_INSTANCE_REF = "clientInstanceRef";
+
+    @Override
+    public Optional<String> getTokenBindingValue(OAuth2AccessTokenReqDTO oAuth2AccessTokenReqDTO) {
+
+        RequestParameter[] parameters = oAuth2AccessTokenReqDTO.getRequestParameters();
+        for (RequestParameter parameter : parameters) {
+            if (CLIENT_INSTANCE_REF.equals(parameter.getKey())
```
Comment (senthalan) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/bindings/impl/ClientInstanceTokenBinder.java:49]: can we verify how easy is to send this parameter in the token request using the SDk ? does SDKs allow to add extra parameter in the token request ?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/AccessTokenIssuer.java (hunk 2 above)
Comment (VivekVinushanth) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/AccessTokenIssuer.java:1080]: We can but we are unable to throw a client exception from the binder itself.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/bindings/impl/ClientInstanceTokenBinder.java (hunk 3 above)
Comment (VivekVinushanth) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/bindings/impl/ClientInstanceTokenBinder.java:49]: AFAIK, our SDKs do not support sending additional parameters. But wondering back-channel grant types like password/token-exchange require an SDK?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/AccessTokenIssuer.java (hunk 1 above)
Comment (VivekVinushanth) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/AccessTokenIssuer.java:1069]: If the client instance has not sent any value, we treat as no binding type (as the binding is set to null by default)

**Review by senthalan — APPROVED** [review]

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/bindings/impl/ClientRequestTokenBinder.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (c) 2022, WSO2 Inc. (http://www.wso2.com).
```
Comment (kayathiri4) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/bindings/impl/ClientRequestTokenBinder.java:2]: Shall we update License header

**Pull request merged**